core/src/ephemeral_server/mod.rs

```diff
@@ -0,0 +1,454 @@
+
+        // Get free port if not already given
+        let port = self.port.unwrap_or_else(get_free_port);
+
+        // Build arg set
+        let mut args = vec![
+            "start".to_owned(),
+            "--port".to_owned(),
+            port.to_string(),
+            "--namespace".to_owned(),
+            self.namespace.clone(),
+            "--ip".to_owned(),
+            self.ip.clone(),
+            "--log-format".to_owned(),
+            self.log.0.clone(),
+            "--log-level".to_owned(),
+            self.log.1.clone(),
+        ];
+        if let Some(db_filename) = &self.db_filename {
+            args.push("--filename".to_owned());
+            args.push(db_filename.clone());
+        } else {
+            args.push("--ephemeral".to_owned());
+        }
+        if !self.ui {
+            args.push("--headless".to_owned());
+        }
+        args.extend(self.extra_args.clone());
+
+        // Start
+        EphemeralServer::start(EphemeralServerConfig {
+            exe_path,
+            port,
+            args,
+            has_test_service: false,
+        })
+        .await
+    }
+}
+
+/// Configuration for the test server.
+#[derive(Debug, Clone, derive_builder::Builder)]
+pub struct TestServerConfig {
+    /// Required path to executable or download info.
+    pub exe: EphemeralExe,
+    /// Port to use or obtains a free one if none given.
+    #[builder(default)]
+    pub port: Option<u16>,
+    /// Additional arguments to the test server.
+    #[builder(default)]
+    pub extra_args: Vec<String>,
+}
+
+impl TestServerConfig {
+    /// Start a test server.
+    pub async fn start_server(&self) -> anyhow::Result<EphemeralServer> {
+        // Get exe path
+        let exe_path = self.exe.get_or_download("temporal-test-server").await?;
+
+        // Get free port if not already given
+        let port = self.port.unwrap_or_else(get_free_port);
+
+        // Build arg set
+        let mut args = vec![port.to_string()];
+        args.extend(self.extra_args.clone());
+
+        // Start
+        EphemeralServer::start(EphemeralServerConfig {
+            exe_path,
+            port,
+            args,
+            has_test_service: true,
+        })
+        .await
+    }
+}
+
+struct EphemeralServerConfig {
+    exe_path: PathBuf,
+    port: u16,
+    args: Vec<String>,
+    has_test_service: bool,
+}
+
+/// Server that will be stopped when dropped.
+pub struct EphemeralServer {
+    /// gRPC target host:port for the server frontend.
+    pub target: String,
+    /// Whether the target implements the gRPC TestService
+    pub has_test_service: bool,
+    _child: tokio::process::Child,
+}
+
+impl EphemeralServer {
+    async fn start(config: EphemeralServerConfig) -> anyhow::Result<EphemeralServer> {
+        // Start process
+        // TODO(cretz): Offer stdio suppression?
+        let child = tokio::process::Command::new(config.exe_path)
+            .args(config.args)
+            .kill_on_drop(true)
```

Comment:
I'm now thinking we need an explicit shutdown that waits for the process to die. People from lang side expect a completed shutdown to mean the port is free.

Reply:
Fixed